### PR TITLE
[Internal] Documentation: Refactors SDK design and component documentation

### PR DESCRIPTION
**Pull request by ealsur**
Refactor of the SDK design document. The previous document had only a big graph with several components but no explanation.

Revamping the entire content with a top-down approach, explaining each component layer with code links

Render preview: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/users/ealsur/diagrams/docs/SdkDesign.md

### CODE REVIEW

**Review by philipthomas-MSFT — DISMISSED** [review]

**Review by kundadebdatta — APPROVED** [review]

File: docs/SdkDesign.md [hunk 1]
```diff
@@ -1,62 +1,173 @@
...
+    IsSuccess -- No --> IsRetriable{{Transient retryable error / reached max latency?}}
+    IsRetriable -- Yes --> HttpTimeoutPolicy
 
 ```
+
+## Direct mode retry layer
+
+Direct connectivity is obtained from the `Microsoft.Azure.Cosmos.Direct` package reference.
+
+> There is an example code branch in the repository for reference purposes that will be linked in this document.
```
Comment (jcocchi) [inline docs/SdkDesign.md:132]: Can we add the link here?

File: docs/SdkDesign.md [hunk 2]
```diff
@@ -1,62 +1,173 @@
...
+
+```
+
+## Consistency (direct mode)
+
+When performing operations through Direct mode, the SDK is involved in checking consistency for Bounded Staleness and Strong accounts. Read requests are handled by the [ConsistencyReader](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/ConsistencyReader.cs) and write requests are handled by the [ConsistencyWriter](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/ConsistencyWriter.cs). The `ConsistencyReader` uses the [QuorumReader](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/QuorumReader.cs) for cases when the consistency is Bounded Staleness or Strong, in which cases, the SDK verifies quorum after performing two requests and comparing the LSNs, if quorum cannot be achieved, the SDK starts what is defined as "barrier requests" to the container waiting for it to achieve quorum. The `ConsistencyWriter` also performs a similar LSN check, after receiving the response from the write, the `GlobalCommittedLSN` and the item `LSN`, if they don't match, barrier requests are also performed.
```
Comment (jcocchi) [inline docs/SdkDesign.md:160]: nit: tried to clean up some sentences, but please make sure the meaning is still correct. 

```suggestion
When performing operations through Direct mode, the SDK is involved in checking consistency for Bounded Staleness and Strong accounts. Read requests are handled by the [ConsistencyReader](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/ConsistencyReader.cs) and write requests are handled by the [ConsistencyWriter](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/ConsistencyWriter.cs). The `ConsistencyReader` uses the [QuorumReader](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/msdata/direct/Microsoft.Azure.Cosmos/src/direct/QuorumReader.cs) when the consistency is Bounded Staleness or Strong to verify quorum after performing two requests and comparing the LSNs. If quorum cannot be achieved, the SDK starts what is defined as "barrier requests" to the container and waits for it to achieve quorum. The `ConsistencyWriter` also performs a similar LSN check after receiving the response from the write, the `GlobalCommittedLSN` and the item `LSN`. If they don't match, barrier requests are also performed.
```

**Review by jcocchi — DISMISSED** [review]
Looks good to me! Added a few nits and questions

File: docs/SdkDesign.md [hunk 3]
```diff
@@ -1,62 +1,173 @@
...
+    InvalidPartitionExceptionRetryHandler <--> PartitionKeyRangeHandler
+    PartitionKeyRangeHandler <--> TransportHandler
+    TransportHandler <--> TransportClient[[Selected Transport]]
+```
+
+## Cross region retries
+
+Any failure response from the Transport that matches the [conditions for cross-regional communication](https://docs.microsoft.com/azure/cosmos-db/troubleshoot-sdk-availability) are handled by the [RetryHandler](../Microsoft.Azure.Cosmos/src/Handler/RetryHandler.cs) through the [ClientRetryPolicy](../Microsoft.Azure.Cosmos/src/ClientRetryPolicy.cs).
+
+* HTTP connection failures or DNS resolution problems (HttpRequestException) - The account information is refreshed, the current region is marked unavailable to be used, and the request is retried on the next available region after account refresh (if the account has multiple regions).
```
[the remark this answers is not in the document]
Comment (ealsur) [reply docs/SdkDesign.md:45]: My intent is to indicate that if there is no other region, then the request fails. If we retry on the next available region and there is no next region, then the retry does not have anywhere to go.

File: docs/SdkDesign.md [hunk 4]
```diff
@@ -1,62 +1,173 @@
...
+* HTTP 503 - The request could not succeed on the region due to repeated TCP connectivity issues, the request is retried on the next preferred region.
+
+## Transport
+
+Once a RequestMessage reaches the [TransportHandler](../Microsoft.Azure.Cosmos/src/Handler/TransportHandler.cs) it will be sent through either the [GatewayStoreModel](../Microsoft.Azure.Cosmos/src/GatewayStoreModel.cs) for HTTP requests and Gateway mode clients or through the ServerStoreModel for clients configured with Direct mode.
+
+Even on clients configured on Direct mode, there can be [HTTP requests that get routed to Gateway](https://docs.microsoft.com/azure/cosmos-db/sql/sql-sdk-connection-modes#direct-mode). The `ConnectionMode` defined in the `CosmosClientOptions` affect data-plane operations (operations related to Items, like CRUD or query over existing Items in a Container) but metadata/control-plane operations (that appear as [MetadataRequests](https://docs.microsoft.com/azure/cosmos-db/monitor-cosmos-db-reference#request-metrics) on Azure Monitor) are sent through HTTP to Gateway.
```
[the remark this answers is not in the document]
Comment (ealsur) [reply docs/SdkDesign.md:55]: FFCF on Gateway is just during preview, so no, I would not include that corner case

File: docs/SdkDesign.md (hunk 1 above)
Comment (ealsur) [reply docs/SdkDesign.md:132]: I am already doing so, the links below are pointing to that branch.

File: docs/SdkDesign.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ealsur) [reply docs/SdkDesign.md]: Added more information on single region

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**